pkg/user/signer.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	timer := time.NewTimer(s.pollTime)
+	for {
+		select {
+		case <-ctx.Done():
+			return &sdktypes.TxResponse{}, ctx.Err()
+		case <-timer.C:
+			resp, err = txClient.GetTx(
+				ctx,
+				&tx.GetTxRequest{
+					Hash: txHash,
+				},
+			)
+
+			if err == nil {
+				return resp.TxResponse, err
+			}
+
+			if !strings.Contains(err.Error(), "not found") {
+				return &sdktypes.TxResponse{}, err
+			}
+		}
+	}
+}
+
+// ChainID returns the chain ID of the signer.
+func (s *Signer) ChainID() string {
+	return s.chainID
+}
+
+// AccountNumber returns the account number of the signer.
+func (s *Signer) AccountNumber() uint64 {
+	return s.accountNumber
+}
+
+// Address returns the address of the signer.
+func (s *Signer) Address() sdktypes.Address {
+	return s.address
+}
+
+func (s *Signer) signTransaction(builder client.TxBuilder) error {
+	signers := builder.GetTx().GetSigners()
+	if len(signers) != 1 {
+		return fmt.Errorf("expected 1 signer, got %d", len(signers))
+	}
+
+	if !s.address.Equals(signers[0]) {
+		return fmt.Errorf("expected signer %s, got %s", s.address.String(), signers[0].String())
+	}
+
+	sequence := s.getSequence()
+
+	// To ensure we have the correct bytes to sign over we produce
+	// a dry run of the signing data
+	draftsigV2 := signing.SignatureV2{
+		PubKey: s.pk,
+		Data: &signing.SingleSignatureData{
+			SignMode:  signing.SignMode_SIGN_MODE_DIRECT,
+			Signature: nil,
+		},
+		Sequence: sequence,
+	}
+
+	err := builder.SetSignatures(draftsigV2)
+	if err != nil {
+		return fmt.Errorf("error setting draft signatures: %w", err)
+	}
+
+	// now we can use the data to produce the signature from each signer
+	signature, err := s.createSignature(builder, sequence)
+	if err != nil {
+		return fmt.Errorf("error creating signature: %w", err)
+	}
+	sigV2 := signing.SignatureV2{
+		PubKey: s.pk,
+		Data: &signing.SingleSignatureData{
+			SignMode:  signing.SignMode_SIGN_MODE_DIRECT,
+			Signature: signature,
+		},
+		Sequence: sequence,
+	}
+
+	err = builder.SetSignatures(sigV2)
+	if err != nil {
+		return fmt.Errorf("error setting signatures: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Signer) createSignature(builder client.TxBuilder, sequence uint64) ([]byte, error) {
+	signerData := authsigning.SignerData{
+		Address:       s.address.String(),
+		ChainID:       s.ChainID(),
+		AccountNumber: s.accountNumber,
+		Sequence:      sequence,
+		PubKey:        s.pk,
+	}
+
+	bytesToSign, err := s.encCfg.TxConfig.SignModeHandler().GetSignBytes(
+		signing.SignMode_SIGN_MODE_DIRECT,
+		signerData,
+		builder.GetTx(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error getting sign bytes: %w", err)
+	}
+
+	signature, _, err := s.keys.SignByAddress(s.address, bytesToSign)
+	if err != nil {
+		return nil, fmt.Errorf("error signing bytes: %w", err)
+	}
+
+	return signature, nil
+}
+
+// NewTxBuilder returns the default sdk Tx builder using the celestia-app encoding config
+func (s *Signer) txBuilder(opts ...TxOption) client.TxBuilder {
```

Comment:
[not blocking] I think the godoc didn't get updated after a function rename

```suggestion
// txBuilder returns the default sdk Tx builder using the celestia-app encoding config
func (s *Signer) txBuilder(opts ...TxOption) client.TxBuilder {
```